src/psyclone/domain/common/transformations/alg_invoke_2_psy_call_trans.py

```diff
@@ -97,29 +108,49 @@ def apply(self, node, options=None):
 
         '''
         self.validate(node, options=options)
-
         node.create_psylayer_symbol_root_names()
+        arguments = self.get_arguments(node, options=options)
+        symbol_table = node.ancestor(Routine).symbol_table
 
-        arguments = []
-        sym_maths = SymbolicMaths.get()
-        for kern in node.children:
-            for arg in kern.children:
-                if isinstance(arg, Literal):
-                    # Literals are not passed by argument.
-                    pass
-                elif isinstance(arg, (Reference, ArrayReference)):
-                    for existing_arg in arguments:
-                        if sym_maths.equal(arg, existing_arg):
+        # Get a unique set of kernel functor symbols for this invoke
+        # if they are explicitly imported.
+        kernel_functor_symbols = set()
+        for kernel_functor in node.children:
+            if kernel_functor.symbol.is_import:
+                kernel_functor_symbols.add(kernel_functor.symbol)
+        # Remove imported symbols as appropriate
+        for kernel_functor_symbol in kernel_functor_symbols:
+            # Is this kernel_functor used in a different invoke?
+            used_elsewhere = False
+            # Search from where the symbol is declared
+            kf_symbol_table = kernel_functor_symbol.find_symbol_table(node)
+            scope_node = kf_symbol_table.node
+            for invoke in scope_node.walk(AlgorithmInvokeCall):
+                if id(invoke) != id(node):
+                    for kernel_functor in invoke.children:
+                        if kernel_functor.symbol == kernel_functor_symbol:
+                            used_elsewhere = True
```

Comment:
I've moved it into another method but am not sure what you mean about returning early as we are iterating over all functors within an AlgorithmInvokeCall. Were you thinking that the method should only deal with one functor? Actually I can't see how we would return early in that case either. Did you mean use return, rather than setting a logical value to break out of the nested loop?

Reply:
The suggested solution works nicely and is much cleaner, thanks! I purposely use `is_import` here as this code is common and `BuiltinFunctor` is an LFRic class. I also think it is more robust to look for imported symbols as we are including what we want, rather than what we don't want.